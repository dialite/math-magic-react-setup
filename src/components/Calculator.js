/* eslint-disable operator-linebreak */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prefer-stateless-function */
import React, { useState } from 'react';
import '../App.css';
import calculate from '../logic/calculate';
import '../logic/operate';

const Calculator = () => {
  const [data, setData] = useState({
    total: 0,
    next: null,
    operation: null,
    err: null,
  });

  const clickHandler = (e) => {
    setData((prevData) => calculate(prevData, e.target.name));
  };

  return (
    <div className="main-container">
      <h2>Lets Do Some Math</h2>
      <div className="calculation-field">
        <h1 className="answer" data-testid="testH2">
          {data.next || data.operation || data.total || data.err || 0}
        </h1>
        <div className="numbers-and-symbols">
          <button
            type="button"
            onClick={clickHandler}
            className="AC size"
            name="AC"
          >
            AC
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="plus-or-minus size"
            name="+/-"
          >
            +/-
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="percentage size"
            name="%"
          >
            %
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="division size orabg-bg"
            name="รท"
          >
            รท
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="seven size"
            name="7"
          >
            7
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="eight size"
            name="8"
          >
            8
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="nine size"
            name="9"
          >
            9
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="cross size orabg-bg"
            name="x"
          >
            x
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="four size"
            name="4"
          >
            4
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="five size"
            name="5"
          >
            5
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="six size"
            name="6"
          >
            6
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="minus size orabg-bg"
            name="-"
          >
            -
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="one size"
            name="1"
          >
            1
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="two size"
            name="2"
          >
            2
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="three size"
            name="3"
          >
            3
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="plus size orabg-bg"
            name="+"
          >
            +
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="zero"
            name="0"
          >
            0
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="dot size"
            name="."
          >
            .
          </button>
          <button
            type="button"
            onClick={clickHandler}
            className="equal size orabg-bg"
            name="="
          >
            =
          </button>
        </div>
      </div>
    </div>
  );
};

export default Calculator;
