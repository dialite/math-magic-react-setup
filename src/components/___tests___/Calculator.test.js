import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, fireEvent } from '@testing-library/react';
// import Calculator from '../Calculator';
import Calculator from "../Calculator";

describe('Test render of Calculator component', () => {
  it('Calculator elements are rendered correctly', () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  it('If buttons 5 and 6 are pressed, Calculator screen should change display from 0 to 56', () => {
    // ARRANGE
    render(<Calculator />);
    const button5 = screen.getByText('5');
    const button6 = screen.getByText('6');
    const resultElement = screen.getByTestId('calculator-screen-id');
    // ACT
    fireEvent.click(button5);
    fireEvent.click(button6);
    // ASSERT
    expect(resultElement).toMatchSnapshot();
    expect(resultElement).toHaveTextContent('56');
  });
  it(`If buttons 5 and + are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to 11`, () => {
    // ARRANGE
    render(<Calculator />);
    const button5 = screen.getByText('5');
    const button6 = screen.getByText('6');
    const additionButton = screen.getByText('+');
    const equalButton = screen.getByText('=');
    const resultElement = screen.getByTestId('calculator-screen-id');
    // ACT
    fireEvent.click(button5);
    fireEvent.click(additionButton);
    fireEvent.click(button6);
    fireEvent.click(equalButton);
    // ASSERT
    expect(resultElement).toMatchSnapshot();
    expect(resultElement).toHaveTextContent('11');
  });
  it(`If buttons 5 and - are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to -1`, () => {
    // ARRANGE
    render(<Calculator />);
    const button5 = screen.getByText('5');
    const button6 = screen.getByText('6');
    const substractionButton = screen.getByText('-');
    const equalButton = screen.getByText('=');
    const resultElement = screen.getByTestId('calculator-screen-id');
    // ACT
    fireEvent.click(button5);
    fireEvent.click(substractionButton);
    fireEvent.click(button6);
    fireEvent.click(equalButton);
    // ASSERT
    expect(resultElement).toMatchSnapshot();
    expect(resultElement).toHaveTextContent('-1');
  });
  it(`If buttons 5 and x are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to 30`, () => {
    // ARRANGE
    render(<Calculator />);
    const button5 = screen.getByText('5');
    const button6 = screen.getByText('6');
    const multiplyButton = screen.getByText('Ã—');
    const equalButton = screen.getByText('=');
    const resultElement = screen.getByTestId('calculator-screen-id');
    // ACT
    fireEvent.click(button5);
    fireEvent.click(multiplyButton);
    fireEvent.click(button6);
    fireEvent.click(equalButton);
    // ASSERT
    expect(resultElement).toMatchSnapshot();
    expect(resultElement).toHaveTextContent('30');
  });
});
