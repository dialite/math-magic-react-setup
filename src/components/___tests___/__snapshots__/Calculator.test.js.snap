// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render of Calculator component Calculator elements are rendered correctly 1`] = `
<div>
  <div
    class="calculator"
  >
    <div
      class="calculator-screen gray"
      data-testid="calculator-screen-id"
    >
      0
    </div>
    <button
      class="operator divide orange"
      type="button"
      value="รท"
    >
      รท
    </button>
    <button
      class="operator multiply orange"
      type="button"
      value="x"
    >
      ร
    </button>
    <button
      class="operator substract orange"
      type="button"
      value="-"
    >
      -
    </button>
    <button
      class="operator add orange"
      type="button"
      value="+"
    >
      +
    </button>
    <button
      class="equal-sign orange"
      type="button"
      value="="
    >
      =
    </button>
    <button
      class="all-clear"
      type="button"
      value="AC"
    >
      AC
    </button>
    <button
      class="invert-sign"
      type="button"
      value="+/-"
    >
      +/-
    </button>
    <button
      class="percentage"
      type="button"
      value="%"
    >
      %
    </button>
    <button
      class="number-key zero"
      type="button"
      value="0"
    >
      0
    </button>
    <button
      class="number-key one"
      type="button"
      value="1"
    >
      1
    </button>
    <button
      class="number-key two"
      type="button"
      value="2"
    >
      2
    </button>
    <button
      class="number-key three"
      type="button"
      value="3"
    >
      3
    </button>
    <button
      class="number-key four"
      type="button"
      value="4"
    >
      4
    </button>
    <button
      class="number-key five"
      type="button"
      value="5"
    >
      5
    </button>
    <button
      class="number-key six"
      type="button"
      value="6"
    >
      6
    </button>
    <button
      class="number-key seven"
      type="button"
      value="7"
    >
      7
    </button>
    <button
      class="number-key eight"
      type="button"
      value="8"
    >
      8
    </button>
    <button
      class="number-key nine"
      type="button"
      value="9"
    >
      9
    </button>
    <button
      class="decimal"
      type="button"
      value="."
    >
      .
    </button>
  </div>
</div>
`;

exports[`Test render of Calculator component If buttons 5 and + are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to 11 1`] = `
<div
  class="calculator-screen gray"
  data-testid="calculator-screen-id"
>
  11
</div>
`;

exports[`Test render of Calculator component If buttons 5 and - are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to -1 1`] = `
<div
  class="calculator-screen gray"
  data-testid="calculator-screen-id"
>
  -1
</div>
`;

exports[`Test render of Calculator component If buttons 5 and 6 are pressed, Calculator screen should change display from 0 to 56 1`] = `
<div
  class="calculator-screen gray"
  data-testid="calculator-screen-id"
>
  56
</div>
`;

exports[`Test render of Calculator component If buttons 5 and x are pressed,
  Then we press 6 button and then =, Calculator screen should change display from 0 to 30 1`] = `
<div
  class="calculator-screen gray"
  data-testid="calculator-screen-id"
>
  30
</div>
`;
